### PR TITLE
Add `RegisterClientWoodTypeMaterialsEvent` to replace `WoodType#register`

### PR DESCRIPTION
**Pull request by Spinoscythe**
`WoodType#register` is not thread-safe, and since we are slowly moving client registrations to their own events, I believe registering wood types should be changed to having its own event as well.

### CODE REVIEW

**Comment by ChampionAsh5357** [comment]
How is the event supposed to be called? Also, why is the justification for the event that the method is not thread-safe? The entry point mentions it's not thread safe and should be called within `FMLClientSetupEvent#enqueueWork`.

Please provide both the reason for the event and implement an entry point for the event to be called. Update all relevant test cases too.

**Comment by Technici4n** [comment]
The general strategy is to move towards registration events for all the different registries in the game. It might not be practical however, but it is what we have done for all the client registries.

**Comment by Spinoscythe** [comment]
> How is the event supposed to be called? Also, why is the justification for the event that the method is not thread-safe? The entry point mentions it's not thread safe and should be called within `FMLClientSetupEvent#enqueueWork`.
> 
> Please provide both the reason for the event and implement an entry point for the event to be called. Update all relevant test cases too.

Fixed!

File: src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypeEvent.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.util.Map;
+import net.minecraft.client.renderer.Sheets;
+import net.minecraft.client.resources.model.Material;
+import net.minecraft.world.level.block.state.properties.WoodType;
+import net.neoforged.bus.api.Event;
+import net.neoforged.fml.event.IModBusEvent;
+import org.jetbrains.annotations.ApiStatus;
+
+public class RegisterWoodTypeEvent extends Event implements IModBusEvent {
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypeEvent.java]: Event name should be `RegisterWoodTypesEvent`.

Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypeEvent.java]: Please give javadoc.

**Comment by Technici4n** [comment]
Wait, this is an event to register wood types _ON THE CLIENT_? This is quite weird...

**Comment by Spinoscythe** [comment]
> Wait, this is an event to register wood types _ON THE CLIENT_? This is quite weird...

yes, it is

File: src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypeEvent.java (hunk 1 above)
Comment (Spinoscythe) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypeEvent.java]: done!

Comment (Spinoscythe) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypeEvent.java]: done!

**Comment by Spinoscythe** [comment]
@Technici4n is the 'needs response' label needed anymore?

**Comment by Technici4n** [comment]
I don't understand why you are commenting "looks like it" on your own PR that you designed to fulfill a specific purpose.

**Comment by Spinoscythe** [comment]
> I don't understand why you are commenting "looks like it" on your own PR that you designed to fulfill a specific purpose.

my bad

**Comment by dhyces** [comment]
It should be renamed to reflect that it isn't for registering a `WoodType` in the sense of `WoodType.register`, but to register a particular `WoodType` with a `Material`. `RegisterWoodTypeMaterialEvent` or some such appropriate name.

File: patches/net/minecraft/client/renderer/Sheets.java.patch [hunk 2]
```diff
@@ -3,29 +3,36 @@
...
  
-     private static Material createSignMaterial(WoodType p_173386_) {
+     public static Material createSignMaterial(WoodType p_173386_) {
 -        return new Material(SIGN_SHEET, new ResourceLocation("entity/signs/" + p_173386_.name()));
 +        ResourceLocation location = new ResourceLocation(p_173386_.name());
 +        return new Material(SIGN_SHEET, new ResourceLocation(location.getNamespace(), "entity/signs/" + location.getPath()));
      }
  
-     private static Material createHangingSignMaterial(WoodType p_251735_) {
+     public static Material createHangingSignMaterial(WoodType p_251735_) {
```
Comment (dhyces) [inline patches/net/minecraft/client/renderer/Sheets.java.patch]: Why are these patched to be public while also using an AT to make them public... Remove these in favor of the AT entries

File: src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypesEvent.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Allows users to register custom {@link WoodType wood types}.
+ *
+ * <p>This event is not {@linkplain ICancellableEvent cancellable}.
+ *
+ * <p>This event is fired on the mod-specific event bus, only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterWoodTypesEvent extends Event implements IModBusEvent {
```
Comment (dhyces) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypesEvent.java]: Rename to reflect client-only intention

**Review by dhyces — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/client/renderer/Sheets.java.patch (hunk 2 above)
Comment (dhyces) [reply patches/net/minecraft/client/renderer/Sheets.java.patch]: as per gigaherz in neocord
> the process always needs to be add ATs -> setup -> edit code -> build and generate patches

if done in the wrong order, patches will generate with ATs applied, which is erroneous

File: src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypesEvent.java (hunk 3 above)
Comment (Spinoscythe) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterWoodTypesEvent.java]: have renamed to different name!

File: patches/net/minecraft/client/renderer/Sheets.java.patch (hunk 2 above)
Comment (Spinoscythe) [reply patches/net/minecraft/client/renderer/Sheets.java.patch]: still results in this happening.

**Comment by Spinoscythe** [comment]
Too niche, hence the closing

**Pull request closed without merging**